[css-ui-3] Add tests for text-overflow

gsnedders opened the pull request:
Originally posted as https://github.com/w3c/csswg-test/pull/1196 by @frivoal on 01 Feb 2017, 15:55 UTC:

> 
> 
> <!-- Reviewable:start -->

<!-- Reviewable:end -->

frivoal committed:
[css-ui-3] Add tests for text-overflow

frivoal committed:
[css-ui-3] Correct spec links

frivoal committed:
[css-ui-3] Change test to fit even when non-overlay scrollbars are used

frivoal committed:
[css-ui-3] Clarify test instructions

frivoal committed:
[css-ui-3] Rework text-overflow tests to make them easier to understand

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-276699046 by @syncbot on 01 Feb 2017, 16:08 UTC:

> ### Automatic validation checks of commit 76bdf9a passed.

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-289191564 by @fantasai on 25 Mar 2017, 06:02 UTC:

> So, my initial set of comments here is that most of your tests are optimized to make a green square at a great cost to understandability. It's definitely nice to share common references. Going a little out of your way to do so is fine. But making a test so convoluted that it is 10x more difficult to understand is not an acceptable cost to doing so. Please re-review your green square tests and rewrite as needed.
> 
> A few comments on the non-square tests:
> * 019 says “are selected” ... there are two things to check here -- one is whether they appear highlighted, the other is whether they will copy when a copy command is initiated. You need to check both, explicitly.
> * 021 needs shorter text, as it currently overflows (due to scrollbars taking up space). I would recommend merely using the word PASS, which is traditional for such things, and using a left margin instead of space characters to ensure that the word PASS is positioned “offscreen”.
> * 018 is a pretty good test, but I'm finding it quite difficult in Firefox to actually select everything in the box. It seems to work if I select from above+left of the top left corner to below+right of the bottom right corner, but as there isn't much space for the cursor there without running into other pieces of the test (like the textarea), I'd suggest increasing the margin/border/padding to make that easier. Insufficient scope in the selection drag results in copying out pieces of the elided text rather than the whole thing. Alternately maybe that qualifies as a fail. I'm not sure.

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-289197465 by @syncbot on 25 Mar 2017, 08:19 UTC:

> ### Automatic validation checks of commit 8a78dae passed.

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-289198561 by @frivoal on 25 Mar 2017, 08:45 UTC:

> @fantasai 
> 
> > Most of your tests are optimized to make a green square at a great cost to understandability. [...] Please re-review your green square tests and rewrite as needed.
> 
> They did not feel that convoluted when I wrote them, but re-reading after having had enough time to forget, I agree that it takes more effort to figure out what they do than it should. I'll review and rework them in that light.
> 
> > 018 is a pretty good test, but [...] Alternately maybe that qualifies as a fail. I'm not sure.
> 
> It qualifies as a fail. You are trying to select the whole elided text to see if it copy&pastes fine, while the spec says that selecting the ellipsis itself should select the elided text. Firefox doesn't do that, which is what makes selecting everything hard. so it's a fail. Do you think it would be enough to clarify the text of the test to make it clear that you're only supposed to try and select the visible "...", or do you think deeper changes to the test are needed?
> 
> > 019 says “are selected” ... there are two things to check here -- one is whether they appear highlighted, the other is whether they will copy when a copy command is initiated. You need to check both, explicitly.
> 
> Hmm, not sure about that. The spec says "If all of the ellipsed text is selected, UAs should show selection of the ellipsis". If the whole text is selected, and the ellipsis is highlighted, and something doesn't copy, that certainly does not sound like a nice behavior, but I don't see which part of css-ui-3 it would violate. If `selectAllChildren()` fails to select everything, that sounds like a spec violation of the Selection API spec, but I am not testing that here. Based on that, I think the test is correct as is, maybe just with a rephrasing from "are selected" to "are highlighted" (or "appear selected" to avoid making assumptions about how UAs show that things are selected). No?
> 
> I suppose I could also write a text that checks that selectAllChildren is unaffected text-overflow, but that feels like something the Selection API spec should test.
> 
> > 021 needs shorter text
> 
> Fixed. It did fit when my OS was configured to use overlay scrollbars, but depending on that was wrong.
> 
> > and using a left margin instead of space characters to ensure that the word PASS is positioned “offscreen”.
> 
> I'm using a monospace font, and sizing things in ch, so the  word PASS is guaranteed to be "offscreen". After that, spaces vs left margin feels subjective, and I find the test easier to read with spaces, so I've left it as is for now. If you have an objective reason (or if you insist that your subjective feeling trumps mine), I'm OK with changing.
>

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-289343300 by @syncbot on 27 Mar 2017, 03:06 UTC:

> ### Automatic validation checks of commit 8ef4dd5 passed.

csswg-test-issue-mover commented:
Originally posted as https://github.com/w3c/csswg-test/pull/1196#issuecomment-289350818 by @frivoal on 27 Mar 2017, 04:22 UTC:

> @fantasai fixes based on the review are done. Ready for next round.
> 
> I've reworded the text instructions in -018 and -019, hopefully the problem you had with them are gone now.
> 
> I've been through the green square tests, and reworked quite a few of them.
> 
> I think the worst offender was by far the first one (-006), so I've made it into a much simpler test with a different reference.
> 
> For most of the other tests, It seemed to me that what made them hard to understand was not so much trying to make them match the standard green square, but the way I was doing it, applying a whole bunch of properties onto a single element, leaving the reviewer with quite a puzzle trying to figure out how it all fits together. I've rewritten them by separating out the various part (the thing that makes a green backdrop, the thing that may turn red on top of it if the test fails...), making it hopefully much clearer and easier to work with, still matching the green square. Hopefully you agree that it is not ok. If not, I'll need more detailed coaching about why it's not.

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/217158824)

# Firefox (nightly channel) #
Testing web-platform-tests at revision cba771875d61cdc819c872457feff5cce5683a9d
Using browser at version BuildID 20170331102157; SourceStamp 8df9fabf2587b7020889755acb9e75b664fe13cf
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>13 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-006.html">/css/css-ui-3/text-overflow-006.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-007.html">/css/css-ui-3/text-overflow-007.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-008.html">/css/css-ui-3/text-overflow-008.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-009.html">/css/css-ui-3/text-overflow-009.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-010.html">/css/css-ui-3/text-overflow-010.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-011.html">/css/css-ui-3/text-overflow-011.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-012.html">/css/css-ui-3/text-overflow-012.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-013.html">/css/css-ui-3/text-overflow-013.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-014.html">/css/css-ui-3/text-overflow-014.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-015.html">/css/css-ui-3/text-overflow-015.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-016.html">/css/css-ui-3/text-overflow-016.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-020.html">/css/css-ui-3/text-overflow-020.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-021.html">/css/css-ui-3/text-overflow-021.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/217158825)

# Chrome (unstable channel) #
Testing web-platform-tests at revision cba771875d61cdc819c872457feff5cce5683a9d
Using browser at version 59.0.3053.3 dev
Starting 10 test iterations
## Unstable results ##

frivoal commented:
@gsnedders This used to pass travis tests in the csswg repo, but it fails here. What's up? I don't understand what it's complaining about.

gsnedders commented:
@frivoal This seems to be a bit of a random failure in the stability checker (note that Travis checks more than it ever did in csswg-test, for example it'll catch flaky tests most of the time), given it just seems to be throwing an exception half the time.

jugglinmike commented:
Hi there!

The confusing report does indeed highlight a bug in this project's automated
validation process. I've filed a patch to fix the bug here:
https://github.com/w3c/web-platform-tests/pull/5330

Note, however, that the bug is only expressed in the presence of instability.
That means that when this patch lands, we should expect the validation to
continue to fail. We won't know exactly why until the bug is fixed, but we can
expect that there is at least one inconsistent test "error" as opposed to an
inconsistent sub test "failure". The distinction here is subtle, but the most
common example is an uncaught JavaScript exception.

w3c-bots commented:
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5285/)

frivoal commented:
Rebased the branch to get the fix for #5330

frivoal commented:
@gsnedders @jugglinmike integrating the fix for #5330 didn't seem to help. Can you help me understand what is still failing?

gsnedders commented:
Probably https://github.com/jugglinmike/chrome-screenshot-race/issues/1.

jugglinmike commented:
> @gsnedders @jugglinmike integrating the fix for #5330 didn't seem to help.
> Can you help me understand what is still failing?

Actually, that patch did fix the infrastructure problem we were seeing here. In
[my previous
comment](https://github.com/w3c/web-platform-tests/pull/5285#issuecomment-291250540),
I wrote:

> That means that when this patch lands, we should expect the validation to
> continue to fail.

The reported failure is related to the test's runtime behavior in the Google
Chrome browser.

> Probably jugglinmike/chrome-screenshot-race#1.

That issue concerns tests that include image assets. I don't think it applies
here.

frivoal commented:
sorry if I'm being dense, but these are simple ref tests that (appear to?) give reliable results when run manually. I don't understand what kind of instability we're talking about.

gsnedders commented:
@frivoal see the issue I linked above; there's something (a bug in Chrome? a bug in ChromeDriver?) that causes screenshots to get taken when not everything is painted

gsnedders commented:
actually probably https://github.com/w3c/wptrunner/issues/241

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240211754)

# Firefox (nightly) #
Testing web-platform-tests at revision 38606e269f3b50dd58c989eb0d107bd6c0a274d2
Using browser at version BuildID 20170606100219; SourceStamp 58ce95bc58ce4ba200413c8bed87786dccf3d105
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>13 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-006.html">/css/css-ui-3/text-overflow-006.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-007.html">/css/css-ui-3/text-overflow-007.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-008.html">/css/css-ui-3/text-overflow-008.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-009.html">/css/css-ui-3/text-overflow-009.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | PASS    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-010.html">/css/css-ui-3/text-overflow-010.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-011.html">/css/css-ui-3/text-overflow-011.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-012.html">/css/css-ui-3/text-overflow-012.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-013.html">/css/css-ui-3/text-overflow-013.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-014.html">/css/css-ui-3/text-overflow-014.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-015.html">/css/css-ui-3/text-overflow-015.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-016.html">/css/css-ui-3/text-overflow-016.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-020.html">/css/css-ui-3/text-overflow-020.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/5285/css/css-ui-3/text-overflow-021.html">/css/css-ui-3/text-overflow-021.html</a></summary>


| Subtest | Results | Messages |
|---------|---------|----------|
|         | FAIL    |          |

</details>
</details>

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240211755)

# Sauce (safari) #
Testing web-platform-tests at revision 38606e269f3b50dd58c989eb0d107bd6c0a274d2
Using browser at version 10.0
Starting 10 test iterations

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240211761)

# Chrome (unstable) #
Testing web-platform-tests at revision 38606e269f3b50dd58c989eb0d107bd6c0a274d2

w3c-bots commented:
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/240211762)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 38606e269f3b50dd58c989eb0d107bd6c0a274d2

fantasai commented on css/css-ui-3/reference/text-overflow-021-ref.html:
```diff
@@ -0,0 +1,23 @@
+</style>
+
+<p>This test passes if the text is in the blue box below is “This test passes”.
+<div id=test>                This test passes</div>
```
This does not appear to render the same as the test file.

frivoal committed:
[css-ui] Make reference match test

frivoal commented:
@fantasai I've fixed the problem you found in https://github.com/w3c/web-platform-tests/pull/5285#pullrequestreview-45889840

frivoal committed:
[css-ui-3] Workaround line height calculations inconsistencies

fantasai commented on css/css-ui-3/text-overflow-013.html:
```diff
@@ -0,0 +1,41 @@
+
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=red></div>
+<div class=test> <span> xx</span> </div>
```
Overall not a bad test, but ...
  - Let's make elided text red and visible so that if it isn't elided the test fails.
  - Ensure that the elided text and the characters before and after it are all encased in the span so that the ellipsis can't be referencing any adjacent direct contents of the block.
  - Also put a line break within the span, both before and after the line being tested, so that the entire line box being tested unquestionably contains only the contents of the span. This way it's clear that it is referencing the baseline of the block even when the block has no direct content on the line. (You can also have elided text on the lines before/after this particular test if you want to test combination lines, but we should have at least one test where there is no direct block content on the line.)

fantasai commented on css/css-ui-3/text-overflow-014.html:
```diff
@@ -0,0 +1,47 @@
+
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=red></div>
+<div class=test> <span>                    </span></div>
```
The elided text here should be x's not spaces, so that they will be visibly red if not elided.

I also think this test would be clearer if written based on the 50px template, since then we just put a little extra text in it instead of fussing with negative sizes, but ymmv.

I do think you can turn the abspos into a red background on the test div, though, right?

frivoal replied:
I am not sure which "50px template" you're referring to. Using a red background on the div here wouldn't work, since that would also be visible through the transparent first character (the space).

fantasai commented on css/css-ui-3/text-overflow-014.html:
```diff
@@ -0,0 +1,47 @@
+	color: red;
+	line-height: 0; /* not clear that this is needed based on specs,
+			   but without it, line height calculations go
+			   wrong (too tall) in some browsers. */
```
I think `color: red; font: 50%/1 serif;` should be fine here, no? Throwing in the font size change seems useful regardless.

fantasai commented on css/css-ui-3/text-overflow-015.html:
```diff
@@ -0,0 +1,29 @@
+  font-size: 20px;
+  font-family: ahem;
+  color: green;
+  background: linear-gradient(red, red) top right / 1em 1em no-repeat, green;
```
Please replace this with swatch-red.png, as we don't want to be depending on css3-images if we don't need to.

fantasai commented on css/css-ui-3/text-overflow-015.html:
```diff
@@ -0,0 +1,29 @@
+<link rel="help" href="http://www.w3.org/TR/css3-ui/#ellipsing-details">
+<link rel="match" href="../reference/ref-filled-green-100px-square.xht">
+<meta name="flags" content="ahem">
+<meta name="assert" content="Ellipsing occurs after relative positioning and other graphical transformations">
```
This is only testing relative positioning, so the assert needs to be that specific.

You also need a test for transforms.

frivoal replied:
Tweaked the test to check for both.

frivoal replied:
Actually, that was a mistake. The spec calls for "and other graphical transformations", but css transforms don't apply to (non-atomic) inlines, and you can only get the ellipsis in a non atomc inline. So the the situation it hints at does not exist, and there's nothing to test for. Reverted.

fantasai commented on css/css-ui-3/text-overflow-016.html:
```diff
@@ -0,0 +1,46 @@
+
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=red></div>
+<div class=green></div>
```
Can you just replace these abspos with `background: green; height: 50px; border-bottom: solid green 50px` on the test div?

fantasai commented on css/css-ui-3/text-overflow-016.html:
```diff
@@ -0,0 +1,46 @@
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=red></div>
+<div class=green></div>
+<div class=test><span></span>filler text to make things overflow</div>
```
This doesn't test whether the ellipsis is clipped. It could be simply shifted to the left so that it overlaps the inline block.

fantasai commented on css/css-ui-3/text-overflow-017.html:
```diff
@@ -0,0 +1,37 @@
+<link rel="help" href="http://www.w3.org/TR/css3-ui/#text-overflow">
+<link rel="match" href="../reference/ref-filled-green-100px-square.xht">
+<meta name="flags" content="ahem interact dom">
+<meta name="assert" content="Ellipsing must not affect dispatching of pointer events.">
```
“must not affect layout nor dispatching of pointer events” is sufficiently vague that I don't know how to review this test. It looks like you're testing that clicking on an ellipsis has the same effect as clicking on the block itself. That's interesting, but what about if the elided text was in a SPAN? Should the event get dispatched to the block or to the SPAN?

frivoal replied:
I'd say it probably should go to the span, but I agree the spec is vague enough that this is probably undefined for now. Would you mind if this was filed as an issue against the text level as something to clarify, and if we kept the tests of this level at what I've written in this test?

fantasai replied:
Do we have interop on this point? Because if we do, we should definitely just fix the spec and the test accordingly.

frivoal replied:
We have interop on this test, all browsers pass it.

However, we do not have full interop on whether or not the event also gets disptached to the span if the elided text is in a span, or to inbetween elements if there's any: Firefox does dispatch the event to all these when clicking on the ellipsis, but other browsers do not, and send it directly to the block element.

Eitherway, I've filled a bug on the spec (l3 & l4) https://github.com/w3c/csswg-drafts/issues/1637.

We may end up clarifying in level 3, but since it's otherwise getting ready to get to PR, at this point I would prefer to limit edits to correcting places where the spec explicitly contradicts (interoperable) behavior that browsers want to keep (if any such thing is found), and leave clarifications of undefined behavior for next level.

Regardless of which way we go, I think this test is correct. We will need more tests when we decide to clarify the sentence (or if we decide that it is already clear and requires one of the behaviors), and I'll be happy to add more tests when that's settled, but this should not block this PR.

frivoal commented on css/css-ui-3/text-overflow-012.html:
```diff
@@ -0,0 +1,75 @@
+document.getElementById("test").style.width = 50 + 50 + (wp + wf) / 2 + "px";
+
+document.documentElement.className = "";
+</script>
```
You'll need to give me a bigger hint than this, because I don't see how I can do that, for the reason I explain in the comment at the beginning of the script.

The size I need to set the element to is font dependent, I don't know the font in advance, and there's no unit corresponding to the aspect of font I'm depending on, so I need to measure. I don't think using "monospace" gives me an escape route, as I do not think I can (reliably) get `1ch` characters here.

frivoal committed:
[css-ui-3] Fixes based on code review

frivoal commented on css/css-ui-3/text-overflow-007.html:
```diff
@@ -0,0 +1,35 @@
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=red></div>
+<div class=test>this line should render to a "t" followed by a "…", both of which are a 1em square in ahem, whithout any gap between them.
+this line also should render to a "t" followed by a "…", both of which are a 1em square in ahem, whithout any gap between them.</div>
```
I thought about this, and kept it the way it is because:
* If you're using AHEM, you're getting a black rectangle instead of a green square. I: 
  a. made it green
  b. added red behind to make failures more obvious
  c. doubled it so that it would be square instead of rectangular.
  a and b seemed to me to be improvements to make life easier for anyone reviewing test results, without actually making the test meaningfully more complicated. 3 may be gratuitous. I thought it reduced the chance of human error when looking at test results after having seen many squares, but I'm happy to get rid of it and add a separate reference if you think it's a net negative.
* If you're not using ahem, the code you suggested doesn't work, and you need to switch from `2.5em` to `2.5ch` and use a monospace font. You also loose the ability when looking at the test manually to tell the difference between the ellipsis being kind of close to the "A", or actually adjacent. Whether AHEM letters touch or not is clearly identifiable.

I think the comment you asked me to add about the `2.5em` width makes the test sufficiently easy to understand, and I'd like to keep it the way it is (modulo getting rid of the doubling to make things square, if you want that gone).

frivoal commented on css/css-ui-3/text-overflow-012.html:
```diff
@@ -0,0 +1,75 @@
+<meta name="flags" content="ahem should">
+<!-- This is SHOULD because although CSS-UI defines "character" as "grapheme cluster" in a non optional manner
+  UAX29 itself (which defines grapheme cluster) only provides a baseline mechnism for determining grapheme clusters,
+  which implementations are expected to tailor and tweak. -->
```
The explanation why this is a should wasn't very good, so I've provided a better one, but I still think we can only justify this as a should level. It would be nice to be stricter, but I don't think the current state of the UAX29 justify that.

frivoal commented on css/css-ui-3/text-overflow-010.html:
```diff
@@ -0,0 +1,36 @@
+
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+<div class=green></div>
+<div class=test>    <span class=ib>X</span>     </div>
```
Done.

frivoal commented:
@fantasai thanks a lot for the review. I've made changes accordingly, except in a few cases documented in answer to your comments. Please let me know how this looks to you now.

frivoal committed:
[css-ui-3] Rescind addition of transforms to text-overflow-015

Transforms do not apply to (non-atomic) inlines, which are required
for text-overflow to apply, so that addition was a mistake.

frivoal committed:
[css-ui] review fixes

frivoal committed:
[css-ui] Added legacy grapheme cluster test

fantasai reviewed (approved):
lgtm

frivoal committed:
[css-ui] add missing file

frivoal committed:
[css-ui] More missing files
